File: android/app/src/main/java/com/stablechannels/app/util/BinarySearch.kt
package com.stablechannels.app.util

import kotlin.math.abs

/**
 * Finds the index of the first element that satisfies the given predicate,
 * assuming elements where predicate returns false precede elements where predicate returns true.
 */
fun <T> List<T>.binarySearchPartitionPoint(predicate: (T) -> Boolean): Int {
    var low = 0
    var high = size

    while (low < high) {
        val mid = low + (high - low) / 2
        if (predicate(this[mid])) {
            high = mid
        } else {
            low = mid + 1
        }
    }

    return low
}

/**
 * Finds the first index where element >= target in a list sorted by selector.
 */
fun <T, K : Comparable<K>> List<T>.lowerBound(target: K, selector: (T) -> K): Int {
    return binarySearchPartitionPoint { selector(it) >= target }
}

/**
 * Finds the first index where element >= target in a naturally sorted list.
 */
fun <T : Comparable<T>> List<T>.lowerBound(target: T): Int {
    return lowerBound(target) { it }
}

/**
 * Finds the first index where element > target in a list sorted by selector.
 */
fun <T, K : Comparable<K>> List<T>.upperBound(target: K, selector: (T) -> K): Int {
    return binarySearchPartitionPoint { selector(it) > target }
}

/**
 * Finds the first index where element > target in a naturally sorted list.
 */
fun <T : Comparable<T>> List<T>.upperBound(target: T): Int {
    return upperBound(target) { it }
}

/**
 * Finds the nearest element to the given target key in a list sorted by the key,
 * using a caller-provided distance metric.
 */
fun <T, K : Comparable<K>, D : Comparable<D>> List<T>.binarySearchNearest(
    target: K,
    selector: (T) -> K,
    distance: (K, K) -> D
): T? {
    if (isEmpty()) return null

    val idx = lowerBound(target, selector)
    if (idx == 0) return first()
    if (idx >= size) return last()

    val prev = this[idx - 1]
    val curr = this[idx]

    return if (distance(selector(prev), target) <= distance(selector(curr), target)) prev else curr
}

/**
 * Convenience overload for Long keys (e.g. timestamps, sats).
 * Distance is computed directly in 64-bit integer arithmetic without floating-point conversions.
 * Note: Assumes difference fits within Long without overflow (e.g. timestamps or satoshi amounts).
 * For arbitrary domains with extreme opposing bounds (e.g. Long.MIN_VALUE and Long.MAX_VALUE),
 * callers should supply a custom distance closure.
 */
@JvmName("binarySearchNearestLong")
fun <T> List<T>.binarySearchNearest(
    target: Long,
    selector: (T) -> Long
): T? {
    return binarySearchNearest(target, selector) { a, b -> if (a >= b) a - b else b - a }
}

/**
 * Convenience overload for Int keys.
 * Distance is computed directly in integer arithmetic without floating-point conversions.
 * Note: Assumes difference fits within Int without overflow.
 */
@JvmName("binarySearchNearestInt")
fun <T> List<T>.binarySearchNearest(
    target: Int,
    selector: (T) -> Int
): T? {
    return binarySearchNearest(target, selector) { a, b -> if (a >= b) a - b else b - a }
}

/**
 * Convenience overload for Double keys.
 */
@JvmName("binarySearchNearestDouble")
fun <T> List<T>.binarySearchNearest(
    target: Double,
    selector: (T) -> Double
): T? {
    return binarySearchNearest(target, selector) { a, b -> abs(a - b) }
}

/**
 * Convenience overload for Float keys.
 */
@JvmName("binarySearchNearestFloat")
fun <T> List<T>.binarySearchNearest(
    target: Float,
    selector: (T) -> Float
): T? {
    return binarySearchNearest(target, selector) { a, b -> abs(a - b) }
}

/**
 * Convenience overload for Date keys.
 */
@JvmName("binarySearchNearestDate")
fun <T> List<T>.binarySearchNearest(
    target: java.util.Date,
    selector: (T) -> java.util.Date
): T? {
    return binarySearchNearest(target, selector) { a, b -> abs(a.time - b.time) }
}

File: ios/StableChannels/StableChannels/Utilities/BinarySearch.swift
import Foundation

extension RandomAccessCollection {
    /// Returns the index of the first element for which the predicate returns false,
    /// assuming the collection is partitioned such that elements where predicate returns
    /// true precede elements where predicate returns false.
    func binarySearchPartitionPoint(where predicate: (Element) -> Bool) -> Index {
        var low = startIndex
        var count = self.count

        while count > 0 {
            let step = count / 2
            let mid = index(low, offsetBy: step)

            if predicate(self[mid]) {
                low = index(after: mid)
                count -= step + 1
            } else {
                count = step
            }
        }

        return low
    }

    /// Finds the first index where element is not ordered before target.
    func lowerBound(
        by areInIncreasingOrder: (Element, Element) -> Bool,
        for target: Element
    ) -> Index {
        binarySearchPartitionPoint { areInIncreasingOrder($0, target) }
    }

    /// Finds the first index where target is ordered before element.
    func upperBound(
        by areInIncreasingOrder: (Element, Element) -> Bool,
        for target: Element
    ) -> Index {
        binarySearchPartitionPoint { !areInIncreasingOrder(target, $0) }
    }

    /// Finds the first index where element key >= target in a collection sorted by key.
    func lowerBound<K: Comparable>(
        target: K,
        keySelector: (Element) -> K
    ) -> Index {
        binarySearchPartitionPoint { keySelector($0) < target }
    }

    /// Finds the first index where element key > target in a collection sorted by key.
    func upperBound<K: Comparable>(
        target: K,
        keySelector: (Element) -> K
    ) -> Index {
        binarySearchPartitionPoint { keySelector($0) <= target }
    }
}

extension RandomAccessCollection where Element: Comparable {
    /// Finds the first index where element >= target.
    func lowerBound(for target: Element) -> Index {
        lowerBound(by: <, for: target)
    }

    /// Finds the first index where element > target.
    func upperBound(for target: Element) -> Index {
        upperBound(by: <, for: target)
    }
}

extension RandomAccessCollection {
    /// Finds the nearest element to the given target key in a collection sorted by the key,
    /// using a caller-provided distance metric.
    func binarySearchNearest<K: Comparable>(
        target: K,
        keySelector: (Element) -> K,
        distance: (K, K) -> some Comparable
    ) -> Element? {
        guard !isEmpty else { return nil }

        let idx = binarySearchPartitionPoint { keySelector($0) < target }

        if idx == startIndex {
            return self[startIndex]
        }
        if idx == endIndex {
            return self[index(before: endIndex)]
        }

        let prevIdx = index(before: idx)
        let prev = self[prevIdx]
        let curr = self[idx]

        return distance(keySelector(prev), target) <= distance(keySelector(curr), target) ? prev : curr
    }

    /// Convenience overload for Date keys.
    func binarySearchNearest(
        target: Date,
        keySelector: (Element) -> Date
    ) -> Element? {
        binarySearchNearest(target: target, keySelector: keySelector) { abs($0.timeIntervalSince($1)) }
    }

    /// Convenience overload for BinaryFloatingPoint keys.
    func binarySearchNearest<K: BinaryFloatingPoint>(
        target: K,
        keySelector: (Element) -> K
    ) -> Element? {
        binarySearchNearest(target: target, keySelector: keySelector) { abs($0 - $1) }
    }

    /// Convenience overload for BinaryInteger keys (e.g. Int, Int64, UInt64 timestamps, block heights, or counts).
    /// Distance is calculated directly in integer space (a >= b ? a - b : b - a) without floating-point conversions.
    /// Note: Assumes values reside within a non-overflowing difference range. For arbitrary wide opposite-signed
    /// domains,
    /// callers should provide a custom distance or widening closure.
    func binarySearchNearest<K: BinaryInteger>(
        target: K,
        keySelector: (Element) -> K
    ) -> Element? {
        binarySearchNearest(target: target, keySelector: keySelector) { $0 >= $1 ? $0 - $1 : $1 - $0 }
    }
}
